examples_utils/custom_ops/custom_ops_utils.py

```diff
@@ -0,0 +1,66 @@
+# Copyright (c) 2022 Graphcore Ltd. All rights reserved.
+
+from time import sleep, time
+
+from cppimport import build_filepath
+import ctypes
+import os
+import shutil
+import sys
+from contextlib import contextmanager
+
+from examples_utils import sdk_version_hash
+
+__all__ = ['load_custom_ops_lib']
+
+
+@contextmanager
+def open_and_delete(path, mode):
+    """Open file and delete it on exit"""
+    try:
+        with open(path, mode) as f:
+            yield f
+    finally:
+        try:
+            os.remove(path)
+        except IOError:
+            sys.stderr.write('Failed to clean up temp file {}'.format(path))
+
+
+def get_binary_path(path_custom_op: str) -> str:
+    return f'{path_custom_op}_{sdk_version_hash()}.so'
+
+
+def load_custom_ops_lib(path_custom_op: str, timeout: int = 5 * 60):
+    """Builds if necessary and loads the custom op binary.
+
+    If the Graphcore SDK version has changed between compilations it automatically recompiles.
+
+    Has safeguards against multiple processes trying to compile at the same time.
+
+    Parameters:
+        path_custom_op (str): path of the custom op C++ file
+        timeout (int): timeout time if cannot obtain lock to compile
+
+    Returns:
+        binary_path: path to binary file
+    """
+    binary_path = get_binary_path(path_custom_op)
+    lock_path = binary_path + '.lock'
+
+    t = time()
+    while not os.path.exists(binary_path) and time() - t < timeout:
+        try:
+            with open_and_delete(lock_path, 'x') as compile_lock:
```

Comment:
You can use a Tempfile here instead https://docs.python.org/3/library/tempfile.html

Reply:
We want the file path to be deterministic and accessible to all processes & hosts trying to compile the op. E.g. if you had multiple poprun instances on different hosts when the op code is on a shared network file system